File: ts/wasm/lib/detailed/complex_operations/comment.ts
import type { TAccountName } from "../hive_apps_operations/factory";
import { operation, comment, asset } from "../../protocol.js";
import { beneficiary_route_type, comment_options, type comment_payout_beneficiaries } from "../../proto/comment_options.js";
import { WaxError } from "../../errors.js";
import { OperationBase, IOperationSink } from "../operation_base.js";
import Long from "long";
import { isNaiAsset } from "../util/asset_util.js";

export enum ECommentFormat {
  HTML = "html",
  MARKDOWN = "markdown",
  MIXED = "markdown+html"
}

export interface ICommentData {
  /**
   * Account name, the author of the commented post or comment.
   * If the operation creates a post, it is empty.
   */
  parentAuthor: TAccountName;
  /**
   * The identifier of the commented post or comment.
   * When a user creates a post, it may contain the identifier of the community (e.g. hive-174695) or main tag (e.g. travel).
   * **It cannot be modified**.
   */
  parentPermlink: string;
  /**
   * Account name, the author of the post or the comment.
   * **It cannot be modified**
   */
  author: TAccountName;
  /**
   * The content of the post or the comment.
   * **It may be modified**.
   */
  body: string;
  /**
   * There is no blockchain validation on json_metadata, but the structure has been established by the community.
   * From the blockchain point of view it is a json file.
   * For the second layer, the following keys may be used:
   * - app, e.g. peakd/2023.2.3
   * - format, e.g. markdown
   * - tags, e.g. photography
   * - users
   * - images
   */
  jsonMetadata: object;
  /**
   * Unique to the author, the identifier of the post or comment.
   * **It cannot be modified**.
   */
  permlink: string;
  /**
   * The title of the submitted post, in case of the comment, is often empty.
   * **It may be modified**.
   */
  title: string;

  /**
   * The format of the comment.
   * Currently supported formats are: html, markdown, markdown+html.
   *
   * @see {@link ECommentFormat}
   */
  format?: ECommentFormat;
  /**
   * The description of the post.
   */
  description?: string;
  /**
   * The links in the post.
   */
  links?: string[];
  /**
   * The images in the post.
   */
  images?: string[];
  /**
   * The alternative author of the post.
   */
  alternativeAuthor?: TAccountName;
  /**
   * The tags of the post.
   */
  tags?: string[];

  /**
   * The beneficiaries of the post including account and the weight.
   * Setting this property will create an additional comment_options operation.
   */
  beneficiaries?: beneficiary_route_type[];

  /**
   * Allow or disallow curation rewards for the post.
   * Setting this property will create an additional comment_options operation.
   *
   * @default true
   */
  allowCurationRewards?: boolean;
  /**
   * Allow or disallow votes for the post.
   * Setting this property will create an additional comment_options operation.
   *
   * @default true
   */
  allowVotes?: boolean;
  /**
   * The percentage of the HBD reward.
   * Setting this property will create an additional comment_options operation.
   *
   * @default 10000
   */
  percentHbd?: number;
  /**
   * The maximum accepted payout for the post.
   * Setting this property will create an additional comment_options operation.
   *
   * @default 1_000_000_000
   */
  maxAcceptedPayout?: asset | Long | string | BigInt | number;
}

export interface IReplyData extends Omit<ICommentData, 'jsonMetadata' | 'permlink' | 'title'> {
  jsonMetadata?: object;
  permlink?: string;
  title?: string;
}

export interface IArticle extends Omit<ICommentData, 'jsonMetadata' | 'permlink' | 'parentAuthor' | 'parentPermlink'> {
  category: string;
  jsonMetadata?: object;
  permlink?: string;
}

class CommentOperation extends OperationBase {
  protected readonly comment: comment;
  private commentOptions?: comment_options;

  protected jsonMetadata: Record<string, any> = {
    format: ECommentFormat.MIXED
  };

  private maxAcceptedPayoutToSet?: asset | Long | string | BigInt | number;

  /**
   * @internal
   */
  public finalize(sink: IOperationSink): Iterable<operation> {
    // Apply cached json metadata before pushing the comment operation
    this.comment.json_metadata = JSON.stringify(this.jsonMetadata);

    const operations: operation[] = [];

    operations.push({ comment: this.comment });

    if(typeof this.commentOptions === "object") {
      if (this.maxAcceptedPayoutToSet !== undefined) {
        let payout: asset;

        if (isNaiAsset(this.maxAcceptedPayoutToSet))
          payout = this.maxAcceptedPayoutToSet as asset;
        else
          payout = sink.api.hbdSatoshis(this.maxAcceptedPayoutToSet as number | string | BigInt | Long);

        this.commentOptions.max_accepted_payout = payout;
      }

      operations.push({ comment_options: this.commentOptions });
    }

    return operations;
  }

  private extendDefaultJsonMetadata(optionalJsonMeta: { app?: string }): void {
    const apps = [ `${process.env.npm_package_name}/${process.env.npm_package_version}` ];
    if(typeof optionalJsonMeta.app === "string")
      apps.unshift(optionalJsonMeta.app);

    Object.assign(this.jsonMetadata, optionalJsonMeta, { app: apps.join(", ") });
  }

  protected constructor(data: ICommentData) {
    super();

    this.comment = comment.fromPartial({
      parent_author: data.parentAuthor,
      parent_permlink: data.parentPermlink,
      author: data.author,
      body: data.body,
      permlink: data.permlink,
      title: data.title
    });

    this.extendDefaultJsonMetadata(data.jsonMetadata);

    if (data.allowCurationRewards !== undefined)
      this.ensureCommentOptionsCreated().allow_curation_rewards = data.allowCurationRewards;

    if (data.allowVotes !== undefined)
      this.ensureCommentOptionsCreated().allow_votes = data.allowVotes;

    if (data.percentHbd !== undefined)
      this.ensureCommentOptionsCreated().percent_hbd = data.percentHbd;

    if (data.maxAcceptedPayout !== undefined) {
      this.ensureCommentOptionsCreated();

      this.maxAcceptedPayoutToSet = data.maxAcceptedPayout;
    }

    if (data.tags !== undefined && data.tags.length > 0) {
      if(this.jsonMetadata.tags === undefined)
        this.jsonMetadata.tags = [];

      this.jsonMetadata.tags = [... new Set([ ...this.jsonMetadata.tags, ...data.tags ])];
    }

    if (data.images !== undefined && data.images.length > 0) {
      if(this.jsonMetadata.image === undefined)
        this.jsonMetadata.image = [];

      this.jsonMetadata.image.push(...data.images);
    }

    if (data.links !== undefined && data.links.length > 0) {
      if(this.jsonMetadata.links === undefined)
        this.jsonMetadata.links = [];

      this.jsonMetadata.links.push(...data.links);
    }

    if (data.alternativeAuthor !== undefined)
      this.jsonMetadata.author = data.alternativeAuthor;

    if (data.description !== undefined)
      this.jsonMetadata.description = data.description;

    if (data.format !== undefined)
      this.jsonMetadata.format = data.format;

    if (data.beneficiaries !== undefined)
      for(const { account, weight } of data.beneficiaries)
        this.addBeneficiary(account, weight);
  }

  private ensureCommentOptionsCreated(): comment_options {
    if(this.commentOptions === undefined) {
      this.maxAcceptedPayoutToSet = 1000000000;

      this.commentOptions = comment_options.fromPartial({
        author: this.comment.author,
        permlink: this.comment.permlink,
        allow_curation_rewards: true,
        allow_votes: true,
        percent_hbd: 10000,
        max_accepted_payout: undefined
      });
    }

    return this.commentOptions;
  }

  /**
   * Assigns given object or sets given value on key in comment meta values
   *
   * @param {string | object} keyOrObject key to set the value on or the entire object of key-value pairs to assign to the json metadata object
   * @param {?any} value value to be set (optional when passing the entire object)
   *
   * @throws {WaxError} if key already exists on the jsonmetadata object
   * @returns {CommentOperation} itself
   */
  public pushMetadataProperty(keyOrObject: string | object, value?: any): CommentOperation {
    const assign = (key: string, value: any) => {
      if(key in this.jsonMetadata)
        throw new WaxError("Key already exists on the json metadata object");

      this.jsonMetadata[key] = value;
    }

    if(typeof keyOrObject === "string")
      assign(keyOrObject, value);
    else
      for(const key in keyOrObject)
        assign(key, keyOrObject[key]);

    return this;
  }

  /**
   * Adds beneficiary account to the comment operation object
   *
   * @param {TAccountName} account beneficiary account
   * @param {number} weight weight of the beneficiary account
   *
   * @returns {CommentOperation} itself
   */
  private addBeneficiary(account: TAccountName, weight: number): CommentOperation {
    this.ensureCommentOptionsCreated();

    let beneficiary: comment_payout_beneficiaries | undefined;

    for(const ext of this.commentOptions!.extensions)
      if("comment_payout_beneficiaries" in ext)
        beneficiary = ext.comment_payout_beneficiaries;


    if(typeof beneficiary === "undefined")
      this.commentOptions!.extensions.push({
        comment_payout_beneficiaries: { beneficiaries: [ { account, weight } ] }
      });
    else
      beneficiary.beneficiaries.push({ account, weight });

    return this;
  }
}

/**
 * Same as the comment base, but requires parentAuthor and parentPermlink to be set
 */
export class ReplyOperation extends CommentOperation {
  public constructor(data: IReplyData) {
    super({
      ...data,
      jsonMetadata: data.jsonMetadata ?? {},
      permlink: data.permlink ?? `re-${data.parentAuthor}-${Date.now()}`,
      title: data.title ?? ""
    });

    if(data.parentAuthor.length === 0)
      throw new WaxError('No parent author specified in the reply operation');

    if(data.parentPermlink.length === 0)
      throw new WaxError('No parent permlink specified in the reply operation');
  }
}

/**
 * Same as the comment base, but requires user to set the category (parent permlink) on the comment
 */
export class BlogPostOperation extends CommentOperation {
  public constructor(data: IArticle) {
    super({
      ...data,
      parentAuthor: '',
      parentPermlink: data.category,
      jsonMetadata: data.jsonMetadata ?? {},
      permlink: data.permlink ?? `${data.author}-${Date.now()}`
    });
  }
}

File: rust/wax/src/base/complex_operations/comment.rs
//! Builders for the `comment_operation` + optional `comment_options_operation`
//! pair, mirroring `ts/wasm/lib/detailed/complex_operations/comment.ts`.
//!
//! Two public builders: [`ReplyOperation`] (reply to an existing post/comment)
//! and [`BlogPostOperation`] (top-level post under a category). Both produce
//! a `comment_operation` and, if the caller set any option-shaped field, a
//! second `comment_options_operation`. The options op is suppressed when the
//! computed options equal the chain defaults — same behaviour as TS's
//! `deepEqual(defaultCommentOptions, this.commentOptions)` check.
//!
//! Differences from the TS port:
//! - `pushMetadataProperty` is not ported as a method; the public
//!   `json_metadata` field covers the same use case. Its entries are merged
//!   into the generated metadata with TS constructor semantics: they
//!   override the `format`/`app` defaults, and the typed fields (`format`,
//!   `tags`, `images`, `links`, `alternative_author`, `description`, `app`)
//!   are applied on top. `app` defaults to `{CARGO_PKG_NAME}/{CARGO_PKG_VERSION}` and
//!   can be overridden via the `app` field or a `json_metadata` entry
//!   (mirrors TS's `jsonMetadata.app`).
//! - The default `comment_options` payload is hard-coded against the
//!   protocol constants rather than fetched through an FFI call to
//!   `cpp_get_default_comment_options_operation`. The values (1_000_000_000
//!   HBD payout, 10_000 bps percent_hbd, both `allow_*` true) are protocol-
//!   stable; this avoids adding bridge surface for a single helper.

use std::time::{SystemTime, UNIX_EPOCH};

use crate::core::proto;
use serde_json::Value;

use crate::WaxError;
use crate::base::constants::{
    DEFAULT_COMMENT_MAX_ACCEPTED_PAYOUT_SATOSHIS, DEFAULT_COMMENT_PERCENT_HBD,
};
use crate::base::foundation::WaxFoundation;
use crate::base::internal::ordered_object::OrderedObject;
use crate::base::models::asset::{AssetName, NaiAsset, NaiAssetConvertible};
use crate::base::models::basic::AccountName;
use crate::base::operation::ComplexOperation;

const APP_TAG: &str =
    concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION"));

/// Represents the wire-form `format` value written to `json_metadata.format`.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub enum CommentFormat {
    Html,
    Markdown,
    /// The format the builders write to `json_metadata.format` when none is
    /// set.
    #[default]
    Mixed,
}

impl CommentFormat {
    fn as_str(self) -> &'static str {
        match self {
            CommentFormat::Html => "html",
            CommentFormat::Markdown => "markdown",
            CommentFormat::Mixed => "markdown+html",
        }
    }
}

/// Represents one beneficiary entry for `comment_options.extensions`, mirroring
/// `proto::BeneficiaryRouteType` without forcing callers to import from
/// `wax::proto`.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct BeneficiaryRoute {
    pub account: AccountName,
    pub weight: u32,
}

/// Represents a reply / comment under an existing post or comment.
/// `parent_author` and `parent_permlink` are required and non-empty; an empty
/// value for either is rejected at `finalize` time.
#[derive(Debug, Clone, Default)]
pub struct ReplyOperation {
    pub parent_author: AccountName,
    pub parent_permlink: String,
    pub author: AccountName,
    pub body: String,
    /// Defaults to `"re-{parent_author}-{millis}"` if `None`.
    pub permlink: Option<String>,
    /// Defaults to `""` if `None`.
    pub title: Option<String>,

    pub format: Option<CommentFormat>,
    pub tags: Vec<String>,
    pub images: Vec<String>,
    pub links: Vec<String>,
    pub alternative_author: Option<AccountName>,
    pub description: Option<String>,
    /// Overrides the `app` tag in `json_metadata`. Defaults to
    /// `{CARGO_PKG_NAME}/{CARGO_PKG_VERSION}` when `None`.
    pub app: Option<String>,
    /// Arbitrary extra `json_metadata` entries, merged in order over the
    /// `format`/`app` defaults before the typed fields above are applied.
    pub json_metadata: Vec<(String, Value)>,

    pub beneficiaries: Vec<BeneficiaryRoute>,
    pub allow_curation_rewards: Option<bool>,
    pub allow_votes: Option<bool>,
    pub percent_hbd: Option<u32>,
    pub max_accepted_payout: Option<NaiAssetConvertible>,
}

/// Represents a top-level blog post under a category (e.g. `"travel"` or a
/// community id like `"hive-174695"`). `parent_author` is always `""`, and
/// `parent_permlink = category`.
#[derive(Debug, Clone, Default)]
pub struct BlogPostOperation {
    pub author: AccountName,
    pub category: String,
    pub title: String,
    pub body: String,
    /// Defaults to `"{author}-{millis}"` if `None`.
    pub permlink: Option<String>,

    pub format: Option<CommentFormat>,
    pub tags: Vec<String>,
    pub images: Vec<String>,
    pub links: Vec<String>,
    pub alternative_author: Option<AccountName>,
    pub description: Option<String>,
    /// Overrides the `app` tag in `json_metadata`. Defaults to
    /// `{CARGO_PKG_NAME}/{CARGO_PKG_VERSION}` when `None`.
    pub app: Option<String>,
    /// Arbitrary extra `json_metadata` entries, merged in order over the
    /// `format`/`app` defaults before the typed fields above are applied.
    pub json_metadata: Vec<(String, Value)>,

    pub beneficiaries: Vec<BeneficiaryRoute>,
    pub allow_curation_rewards: Option<bool>,
    pub allow_votes: Option<bool>,
    pub percent_hbd: Option<u32>,
    pub max_accepted_payout: Option<NaiAssetConvertible>,
}

/// Internal merged representation used by both builders' `finalize` paths.
/// Carries the fully-resolved `parent_*` / `permlink` / `title` values so
/// the shared finalize logic doesn't have to know which builder produced it.
struct CommentInputs {
    parent_author: String,
    parent_permlink: String,
    author: String,
    permlink: String,
    title: String,
    body: String,
    format: Option<CommentFormat>,
    tags: Vec<String>,
    images: Vec<String>,
    links: Vec<String>,
    alternative_author: Option<AccountName>,
    description: Option<String>,
    app: Option<String>,
    json_metadata: Vec<(String, Value)>,
    beneficiaries: Vec<BeneficiaryRoute>,
    allow_curation_rewards: Option<bool>,
    allow_votes: Option<bool>,
    percent_hbd: Option<u32>,
    max_accepted_payout: Option<NaiAssetConvertible>,
}

impl ComplexOperation for ReplyOperation {
    fn finalize(
        self,
        foundation: &WaxFoundation,
    ) -> Result<Vec<proto::Operation>, WaxError> {
        let this = self;
        if this.parent_author.is_empty() {
            return Err(WaxError::new(
                "No parent author specified in the reply operation",
            ));
        }
        if this.parent_permlink.is_empty() {
            return Err(WaxError::new(
                "No parent permlink specified in the reply operation",
            ));
        }

        let permlink = this.permlink.unwrap_or_else(|| {
            format!("re-{}-{}", this.parent_author, now_millis())
        });

        let inputs = CommentInputs {
            parent_author: this.parent_author,
            parent_permlink: this.parent_permlink,
            author: this.author,
            permlink,
            title: this.title.unwrap_or_default(),
            body: this.body,
            format: this.format,
            tags: this.tags,
            images: this.images,
            links: this.links,
            alternative_author: this.alternative_author,
            description: this.description,
            app: this.app,
            json_metadata: this.json_metadata,
            beneficiaries: this.beneficiaries,
            allow_curation_rewards: this.allow_curation_rewards,
            allow_votes: this.allow_votes,
            percent_hbd: this.percent_hbd,
            max_accepted_payout: this.max_accepted_payout,
        };
        inputs.into_operations(foundation)
    }
}

impl ComplexOperation for BlogPostOperation {
    fn finalize(
        self,
        foundation: &WaxFoundation,
    ) -> Result<Vec<proto::Operation>, WaxError> {
        let this = self;
        let permlink = this
            .permlink
            .unwrap_or_else(|| format!("{}-{}", this.author, now_millis()));

        let inputs = CommentInputs {
            parent_author: String::new(),
            parent_permlink: this.category,
            author: this.author,
            permlink,
            title: this.title,
            body: this.body,
            format: this.format,
            tags: this.tags,
            images: this.images,
            links: this.links,
            alternative_author: this.alternative_author,
            description: this.description,
            app: this.app,
            json_metadata: this.json_metadata,
            beneficiaries: this.beneficiaries,
            allow_curation_rewards: this.allow_curation_rewards,
            allow_votes: this.allow_votes,
            percent_hbd: this.percent_hbd,
            max_accepted_payout: this.max_accepted_payout,
        };
        inputs.into_operations(foundation)
    }
}

impl CommentInputs {
    /// `true` iff the caller touched any option-shaped field. Triggers
    /// lazy creation of the `comment_options_operation` per TS semantics.
    fn has_options_input(&self) -> bool {
        self.allow_curation_rewards.is_some()
            || self.allow_votes.is_some()
            || self.percent_hbd.is_some()
            || self.max_accepted_payout.is_some()
            || !self.beneficiaries.is_empty()
    }

    fn into_operations(
        self,
        foundation: &WaxFoundation,
    ) -> Result<Vec<proto::Operation>, WaxError> {
        // Snapshot the options-trigger before consuming `self` field-by-field
        // below — `has_options_input` needs `&self` and would conflict with
        // the partial moves into `proto::Comment`.
        let has_options = self.has_options_input();

        let json_metadata = self.build_json_metadata()?;

        let comment_op = proto::Comment {
            parent_author: self.parent_author,
            parent_permlink: self.parent_permlink,
            author: self.author.clone(),
            permlink: self.permlink.clone(),
            title: self.title,
            body: self.body,
            json_metadata,
        };

        let mut ops = Vec::with_capacity(2);
        ops.push(proto::Operation {
            value: Some(proto::operation::Value::CommentOperation(comment_op)),
        });

        if has_options {
            let default = foundation
                .default_comment_options(&self.author, &self.permlink)?;
            let computed = build_comment_options(
                foundation,
                &self.author,
                &self.permlink,
                self.allow_curation_rewards,
                self.allow_votes,
                self.percent_hbd,
                self.max_accepted_payout,
                &self.beneficiaries,
            )?;

            if computed != default {
                ops.push(proto::Operation {
                    value: Some(
                        proto::operation::Value::CommentOptionsOperation(
                            computed,
                        ),
                    ),
                });
            }
        }

        Ok(ops)
    }

    /// Builds the serialized `json_metadata` string, replicating the TS
    /// constructor sequence: the `format` default, then the user-supplied
    /// `json_metadata` entries (`Object.assign` semantics), then `app`,
    /// then the typed fields on top.
    fn build_json_metadata(&self) -> Result<String, WaxError> {
        let mut metadata = OrderedObject(Vec::new());
        metadata.set("format", Value::from(CommentFormat::default().as_str()));

        for (key, value) in &self.json_metadata {
            metadata.set(key, value.clone());
        }

        let app = match (&self.app, metadata.get("app")) {
            (Some(app), _) => Value::from(app.as_str()),
            (None, Some(app)) => app.clone(),
            (None, None) => Value::from(APP_TAG),
        };
        metadata.set("app", app);

        if !self.tags.is_empty() {
            let combined = metadata_array(&metadata, "tags")?
                .into_iter()
                .chain(self.tags.iter().map(|t| Value::from(t.as_str())));
            metadata.set(
                "tags",
                Value::Array(deduplicate_preserving_order(combined)),
            );
        }

        if !self.images.is_empty() {
            let mut image = metadata_array(&metadata, "image")?;
            image.extend(self.images.iter().map(|i| Value::from(i.as_str())));
            metadata.set("image", Value::Array(image));
        }

        if !self.links.is_empty() {
            let mut links = metadata_array(&metadata, "links")?;
            links.extend(self.links.iter().map(|l| Value::from(l.as_str())));
            metadata.set("links", Value::Array(links));
        }

        if let Some(author) = &self.alternative_author {
            metadata.set("author", Value::from(author.as_str()));
        }

        if let Some(description) = &self.description {
            metadata.set("description", Value::from(description.as_str()));
        }

        if let Some(format) = self.format {
            metadata.set("format", Value::from(format.as_str()));
        }

        serde_json::to_string(&metadata).map_err(|e| {
            WaxError::new(format!(
                "failed to serialize comment json_metadata: {e}"
            ))
        })
    }
}

fn now_millis() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .expect("system clock before unix epoch")
        .as_millis() as u64
}

/// Reads a metadata key that the typed fields extend (`tags`, `image`,
/// `links`), defaulting to an empty array when the key is absent.
/// NOTE: TS throws a `TypeError` when spreading a non-iterable
/// user-supplied value here; surfaced as a [`WaxError`] instead.
fn metadata_array(
    metadata: &OrderedObject,
    key: &str,
) -> Result<Vec<Value>, WaxError> {
    match metadata.get(key) {
        None => Ok(Vec::new()),
        Some(Value::Array(items)) => Ok(items.clone()),
        Some(_) => Err(WaxError::new(format!(
            "comment json_metadata key `{key}` must be an array"
        ))),
    }
}

/// First-occurrence-wins dedupe preserving order. Mirrors TS
/// `[...new Set([...existing, ...data.tags])]`.
fn deduplicate_preserving_order(
    items: impl Iterator<Item = Value>,
) -> Vec<Value> {
    let mut out: Vec<Value> = Vec::new();
    for item in items {
        if !out.contains(&item) {
            out.push(item);
        }
    }

    out
}

#[allow(clippy::too_many_arguments)]
fn build_comment_options(
    foundation: &WaxFoundation,
    author: &str,
    permlink: &str,
    allow_curation_rewards: Option<bool>,
    allow_votes: Option<bool>,
    percent_hbd: Option<u32>,
    max_accepted_payout: Option<NaiAssetConvertible>,
    beneficiaries: &[BeneficiaryRoute],
) -> Result<proto::CommentOptions, WaxError> {
    let max_accepted_payout: NaiAsset = match max_accepted_payout {
        Some(asset) => foundation
            .create_asset_with_required_symbol(AssetName::Hbd, asset)?,
        None => foundation
            .hbd_satoshis(DEFAULT_COMMENT_MAX_ACCEPTED_PAYOUT_SATOSHIS)?,
    };

    let extensions = if beneficiaries.is_empty() {
        Vec::new()
    } else {
        vec![proto::CommentOptionsExtension {
            value: Some(
                proto::comment_options_extension::Value::CommentPayoutBeneficiaries(
                    proto::CommentPayoutBeneficiaries {
                        beneficiaries: beneficiaries
                            .iter()
                            .map(|b| proto::BeneficiaryRouteType {
                                account: b.account.clone(),
                                weight: b.weight,
                            })
                            .collect(),
                    },
                ),
            ),
        }]
    };

    Ok(proto::CommentOptions {
        author: author.to_string(),
        permlink: permlink.to_string(),
        max_accepted_payout,
        percent_hbd: percent_hbd.unwrap_or(DEFAULT_COMMENT_PERCENT_HBD),
        allow_votes: allow_votes.unwrap_or(true),
        allow_curation_rewards: allow_curation_rewards.unwrap_or(true),
        extensions,
    })
}
